# Libraries and loading the dataset
In this first part of the examples we provide we will look at how to load in the dataset, and how to do some basic
manipulation of the data. In order to run the code in this notebook you will need to install, besides Jupyter, the
[Pandas](https://pandas.pydata.org/docs/getting_started/index.html#installation) package.

For more info on how to set up the Python environment for these notebooks, see README.md.

## Getting an overview of the data
Whenever you get a dataset you will want to inspect it and see what variables (columns) you have and what these values
look like. One way of doing this is to use the package [Pandas](https://pandas.pydata.org/docs/getting_started/overview.html),
a package that allows for easy manipulation of the data table (or dataframe, as it is called in Pandas).

For a quick overview of Pandas' capabilities, check out [this](https://pandas.pydata.org/docs/user_guide/10min.html#min)
link.

In [58]:
import pandas as pd

owid = pd.read_csv(r"../data/owid-covid-data.csv")

# Checking table columns and data points
owid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83862 entries, 0 to 83861
Data columns (total 59 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   iso_code                               83862 non-null  object 
 1   continent                              79805 non-null  object 
 2   location                               83862 non-null  object 
 3   date                                   83862 non-null  object 
 4   total_cases                            81878 non-null  float64
 5   new_cases                              81876 non-null  float64
 6   new_cases_smoothed                     80875 non-null  float64
 7   total_deaths                           72262 non-null  float64
 8   new_deaths                             72420 non-null  float64
 9   new_deaths_smoothed                    80875 non-null  float64
 10  total_cases_per_million                81435 non-null  float64
 11  ne

In the above output table you can see the indices of the columns (using standard 0-indexing) as well as the human-readable
name and information about what the data in those columns is. For example: at index 0 we find the `iso_index` this is a
shorthand code for countries (see [here](https://www.iso.org/obp/ui/#search/code/)).

We can also see that there are 83862 'non-nullable' rows for that column. This means non-empty cells in that column. Since
dataframes can be sparsely populated, some rows can have empty column values (e.g. a row can miss the value for `new_tests`
but still have other values).

Lastly, we can see the Python data type for that column. Most will either be floating point numbers or unnamed objects -
meaning they contain multiple values within a single data point.

Note that, in this example, we're only loading in a single dataset. We provided several for you to analyse. Please check
the `data` directory for more data files.

## Fixing the date column
We can see that `date` column is not of the `datetime` object type, a commonly used format for timestamps in Python. It
will be useful to use this column as the x-axis for various plots and analyses, but most of those techniques assume that
the timestamps are in the aforementioned `datetime` format.

This means that if we want to use the `date` column effectively, we will have to alter the data type to `datetime`. The
following code block will show you how to do that. It is recommended to do this for your own analysis too!

In [62]:
# owid['date'] = pd.to_datetime(owid['date'], format = '%Y-%m-%d')
owid['date'].head()

0    2020-02-24
1    2020-02-25
2    2020-02-26
3    2020-02-27
4    2020-02-28
Name: date, dtype: object

In [69]:
owid['date_str'] = [''.join(item.split('-')) for item in owid['date']]
owid['countrycode_date_str'] = [';'.join(item) for item in owid[['iso_code', 'date_str']].values]
owid = owid.set_index('countrycode_date_str')
# print(owid.shape)
# len(owid['countrycode_date_str'].unique())
owid.head()

iso_code continent     location        date  total_cases  \
countrycode_date_str                                                            
AFG;20200224              AFG      Asia  Afghanistan  2020-02-24          1.0   
AFG;20200225              AFG      Asia  Afghanistan  2020-02-25          1.0   
AFG;20200226              AFG      Asia  Afghanistan  2020-02-26          1.0   
AFG;20200227              AFG      Asia  Afghanistan  2020-02-27          1.0   
AFG;20200228              AFG      Asia  Afghanistan  2020-02-28          1.0   

                      new_cases  new_cases_smoothed  total_deaths  new_deaths  \
countrycode_date_str                                                            
AFG;20200224                1.0                 NaN           NaN         NaN   
AFG;20200225                0.0                 NaN           NaN         NaN   
AFG;20200226                0.0                 NaN           NaN         NaN   
AFG;20200227                0.0                 NaN           NaN         NaN   
AFG;20200228                0.0                 NaN           NaN         NaN   

                      new_deaths_smoothed  ...  extreme_poverty  \
countrycode_date_str                       ...                    
AFG;20200224                          NaN  ...              NaN   
AFG;20200225                          NaN  ...              NaN   
AFG;20200226                          NaN  ...              NaN   
AFG;20200227                          NaN  ...              NaN   
AFG;20200228                          NaN  ...              NaN   

                      cardiovasc_death_rate  diabetes_prevalence  \
countrycode_date_str                                               
AFG;20200224                        597.029                 9.59   
AFG;20200225                        597.029                 9.59   
AFG;20200226                        597.029                 9.59   
AFG;20200227                        597.029                 9.59   
AFG;20200228                        597.029                 9.59   

                      female_smokers  male_smokers  handwashing_facilities  \
countrycode_date_str                                                         
AFG;20200224                     NaN           NaN                  37.746   
AFG;20200225                     NaN           NaN                  37.746   
AFG;20200226                     NaN           NaN                  37.746   
AFG;20200227                     NaN           NaN                  37.746   
AFG;20200228                     NaN           NaN                  37.746   

                      hospital_beds_per_thousand  life_expectancy  \
countrycode_date_str                                                
AFG;20200224                                 0.5            64.83   
AFG;20200225                                 0.5            64.83   
AFG;20200226                                 0.5            64.83   
AFG;20200227                                 0.5            64.83   
AFG;20200228                                 0.5            64.83   

                      human_development_index  date_str  
countrycode_date_str                                     
AFG;20200224                            0.511  20200224  
AFG;20200225                            0.511  20200225  
AFG;20200226                            0.511  20200226  
AFG;20200227                            0.511  20200227  
AFG;20200228                            0.511  20200228  

[5 rows x 60 columns]

Your dataset is now ready for your analysis! You may want to sort your data set by date in order to more intuitively see
changes. You can do so doing:

In [92]:
groupedby_date_countrycode = oxcgrt_latest.groupby(['countrycode_date_str']).agg({'Date':'count',
                                                                                  'RegionName':lambda x:list(set(x)),
                                                                                  'CountryCode':lambda x:list(set(x))[0]})
big_countries = set(groupedby_date_countrycode['CountryCode'][groupedby_date_countrycode['Date']>1].value_counts().index.values)

In [78]:
oxcgrt_latest = pd.read_csv('../data/OxCGRT_latest.csv', index_col='CountryName')
oxcgrt_latest['countrycode_date_str'] = [';'.join([str(ele) for ele in item]) for item in oxcgrt_latest[['CountryCode', 'Date']].values]
print(oxcgrt_latest.shape, len(oxcgrt_latest['countrycode_date_str'].unique()))
# oxcgrt_latest = oxcgrt_latest.set_index('countrycode_date_str')
oxcgrt_latest.head()

(135161, 51) 89466


CountryCode RegionName RegionCode Jurisdiction      Date  \
CountryName                                                            
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200101   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200102   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200103   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200104   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200105   

             C1_School closing  C1_Flag  C2_Workplace closing  C2_Flag  \
CountryName                                                              
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   

             C3_Cancel public events  ...  StringencyIndexForDisplay  \
CountryName                           ...                              
Aruba                            0.0  ...                        0.0   
Aruba                            0.0  ...                        0.0   
Aruba                            0.0  ...                        0.0   
Aruba                            0.0  ...                        0.0   
Aruba                            0.0  ...                        0.0   

             StringencyLegacyIndex  StringencyLegacyIndexForDisplay  \
CountryName                                                           
Aruba                          0.0                              0.0   
Aruba                          0.0                              0.0   
Aruba                          0.0                              0.0   
Aruba                          0.0                              0.0   
Aruba                          0.0                              0.0   

             GovernmentResponseIndex  GovernmentResponseIndexForDisplay  \
CountryName                                                               
Aruba                            0.0                                0.0   
Aruba                            0.0                                0.0   
Aruba                            0.0                                0.0   
Aruba                            0.0                                0.0   
Aruba                            0.0                                0.0   

             ContainmentHealthIndex  ContainmentHealthIndexForDisplay  \
CountryName                                                             
Aruba                           0.0                               0.0   
Aruba                           0.0                               0.0   
Aruba                           0.0                               0.0   
Aruba                           0.0                               0.0   
Aruba                           0.0                               0.0   

             EconomicSupportIndex  EconomicSupportIndexForDisplay  \
CountryName                                                         
Aruba                         0.0                             0.0   
Aruba                         0.0                             0.0   
Aruba                         0.0                             0.0   
Aruba                         0.0                             0.0   
Aruba                         0.0                             0.0   

             countrycode_date_str  
CountryName                        
Aruba                ABW;20200101  
Aruba                ABW;20200102  
Aruba                ABW;20200103  
Aruba                ABW;20200104  
Aruba                ABW;20200105  

[5 rows x 51 columns]

In [94]:
# combine except those big countries
print('Big countries are:', big_countries)
isBigCountry = lambda x:True if x in big_countries else False
oxcgrt_latest['isBig'] = [isBigCountry(code) for code in oxcgrt_latest['CountryCode']]
print(oxcgrt_latest['isBig'].value_counts())

smallcountries_oxcgrt = oxcgrt_latest[oxcgrt_latest['isBig']==False].copy()
bigcountries_oxcgrt = oxcgrt_latest[oxcgrt_latest['isBig']==False].copy()

Big countries are: {'USA', 'BRA', 'CAN', 'GBR'}
False    87542
True     47619
Name: isBig, dtype: int64


In [97]:
assert smallcountries_oxcgrt.shape[0] == len(smallcountries_oxcgrt['countrycode_date_str'].unique())

In [99]:
owid['isBig'] = [isBigCountry(code) for code in owid['iso_code']]
smallcountries_owid = owid[owid['isBig']==False]
bigcountries_owid = owid[owid['isBig']]

In [101]:
# combine those small countries
combined_smallcountries = pd.concat([smallcountries_owid, 
                                     smallcountries_oxcgrt.set_index('countrycode_date_str')], axis=1, join='inner')
print(combined_smallcountries.shape)

(70938, 112)


In [104]:
# combine big countries??
bigcountries_oxcgrt.head()

CountryCode RegionName RegionCode Jurisdiction      Date  \
CountryName                                                            
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200101   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200102   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200103   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200104   
Aruba               ABW        NaN        NaN    NAT_TOTAL  20200105   

             C1_School closing  C1_Flag  C2_Workplace closing  C2_Flag  \
CountryName                                                              
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   
Aruba                      0.0      NaN                   0.0      NaN   

             C3_Cancel public events  ...  StringencyLegacyIndex  \
CountryName                           ...                          
Aruba                            0.0  ...                    0.0   
Aruba                            0.0  ...                    0.0   
Aruba                            0.0  ...                    0.0   
Aruba                            0.0  ...                    0.0   
Aruba                            0.0  ...                    0.0   

             StringencyLegacyIndexForDisplay  GovernmentResponseIndex  \
CountryName                                                             
Aruba                                    0.0                      0.0   
Aruba                                    0.0                      0.0   
Aruba                                    0.0                      0.0   
Aruba                                    0.0                      0.0   
Aruba                                    0.0                      0.0   

             GovernmentResponseIndexForDisplay  ContainmentHealthIndex  \
CountryName                                                              
Aruba                                      0.0                     0.0   
Aruba                                      0.0                     0.0   
Aruba                                      0.0                     0.0   
Aruba                                      0.0                     0.0   
Aruba                                      0.0                     0.0   

             ContainmentHealthIndexForDisplay  EconomicSupportIndex  \
CountryName                                                           
Aruba                                     0.0                   0.0   
Aruba                                     0.0                   0.0   
Aruba                                     0.0                   0.0   
Aruba                                     0.0                   0.0   
Aruba                                     0.0                   0.0   

             EconomicSupportIndexForDisplay  countrycode_date_str  isBig  
CountryName                                                               
Aruba                                   0.0          ABW;20200101  False  
Aruba                                   0.0          ABW;20200102  False  
Aruba                                   0.0          ABW;20200103  False  
Aruba                                   0.0          ABW;20200104  False  
Aruba                                   0.0          ABW;20200105  False  

[5 rows x 52 columns]

## Taking a look at the dataframe
The next notebook will deal more with the topic of data visualisation and exploring the data set, but for now you can
visualize (part of) your dataframe by using the following commands:

In [4]:
# Show first N (e.g. 10) rows of the dataframe
data.head(10)

iso_code      continent   location       date  total_cases  new_cases  \
3009       ARG  South America  Argentina 2020-01-01          NaN        NaN   
48722      MEX  North America     Mexico 2020-01-01          NaN        NaN   
48723      MEX  North America     Mexico 2020-01-02          NaN        NaN   
3010       ARG  South America  Argentina 2020-01-02          NaN        NaN   
3011       ARG  South America  Argentina 2020-01-03          NaN        NaN   
48724      MEX  North America     Mexico 2020-01-03          NaN        NaN   
3012       ARG  South America  Argentina 2020-01-04          NaN        NaN   
75171      THA           Asia   Thailand 2020-01-04          NaN        NaN   
48725      MEX  North America     Mexico 2020-01-04          NaN        NaN   
75172      THA           Asia   Thailand 2020-01-05          NaN        NaN   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
3009                  NaN           NaN         NaN                  NaN  ...   
48722                 NaN           NaN         NaN                  NaN  ...   
48723                 NaN           NaN         NaN                  NaN  ...   
3010                  NaN           NaN         NaN                  NaN  ...   
3011                  NaN           NaN         NaN                  NaN  ...   
48724                 NaN           NaN         NaN                  NaN  ...   
3012                  NaN           NaN         NaN                  NaN  ...   
75171                 NaN           NaN         NaN                  NaN  ...   
48725                 NaN           NaN         NaN                  NaN  ...   
75172                 NaN           NaN         NaN                  NaN  ...   

       gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
3009        18933.907              0.6                191.032   
48722       17336.469              2.5                152.783   
48723       17336.469              2.5                152.783   
3010        18933.907              0.6                191.032   
3011        18933.907              0.6                191.032   
48724       17336.469              2.5                152.783   
3012        18933.907              0.6                191.032   
75171       16277.671              0.1                109.861   
48725       17336.469              2.5                152.783   
75172       16277.671              0.1                109.861   

       diabetes_prevalence  female_smokers  male_smokers  \
3009                  5.50            16.2          27.7   
48722                13.06             6.9          21.4   
48723                13.06             6.9          21.4   
3010                  5.50            16.2          27.7   
3011                  5.50            16.2          27.7   
48724                13.06             6.9          21.4   
3012                  5.50            16.2          27.7   
75171                 7.04             1.9          38.8   
48725                13.06             6.9          21.4   
75172                 7.04             1.9          38.8   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
3009                      NaN                        5.00            76.67   
48722                  87.847                        1.38            75.05   
48723                  87.847                        1.38            75.05   
3010                      NaN                        5.00            76.67   
3011                      NaN                        5.00            76.67   
48724                  87.847                        1.38            75.05   
3012                      NaN                        5.00            76.67   
75171                  90.670                        2.10            77.15   
48725                  87.847                        1.38            75.05   
75172                  90.670                        2.10            77.15   

       h

In [5]:
# Tail will show the last N results of the dataframe.
data.tail(10)

iso_code      continent  location       date  total_cases  new_cases  \
46479       MYS           Asia  Malaysia 2021-04-24     390252.0     2717.0   
41871       LAO           Asia      Laos 2021-04-24        247.0       88.0   
59539       PRY  South America  Paraguay 2021-04-24     265296.0     2162.0   
3935        ABW  North America     Aruba 2021-04-24          NaN        NaN   
57045  OWID_OCE            NaN   Oceania 2021-04-24      42985.0      276.0   
75170       TZA         Africa  Tanzania 2021-04-24        509.0        0.0   
28257       GAB         Africa     Gabon 2021-04-24      22433.0        0.0   
80366       URY  South America   Uruguay 2021-04-24     182326.0     2789.0   
21298       DJI         Africa  Djibouti 2021-04-24      10746.0        8.0   
83861       ZWE         Africa  Zimbabwe 2021-04-24      38064.0       19.0   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
46479            2484.714        1426.0        11.0                8.000  ...   
41871              27.000           NaN         NaN                0.000  ...   
59539            2418.857        5802.0        87.0               78.714  ...   
3935                  NaN           NaN         NaN                  NaN  ...   
57045             203.571        1041.0         3.0                3.000  ...   
75170               0.000          21.0         0.0                0.000  ...   
28257              82.143         138.0         0.0                0.714  ...   
80366            2846.571        2283.0        56.0               62.143  ...   
21298              47.714         132.0         0.0                2.571  ...   
83861              52.143        1556.0         0.0                0.571  ...   

       gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
46479       26808.164              0.1                260.942   
41871        6397.360             22.7                368.111   
59539        8827.010              1.7                199.128   
3935        35973.781              NaN                    NaN   
57045             NaN              NaN                    NaN   
75170        2683.304             49.1                217.288   
28257       16562.413              3.4                259.967   
80366       20551.409              0.1                160.708   
21298        2705.406             22.5                258.037   
83861        1899.775             21.4                307.846   

       diabetes_prevalence  female_smokers  male_smokers  \
46479                16.74             1.0          42.4   
41871                 4.00             7.3          51.2   
59539                 8.27             5.0          21.6   
3935                 11.62             NaN           NaN   
57045                  NaN             NaN           NaN   
75170                 5.75             3.3          26.7   
28257                 7.20             NaN           NaN   
80366                 6.93            14.0          19.9   
21298                 6.05             1.7          24.5   
83861                 1.82             1.6          30.7   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
46479                     NaN                         1.9            76.16   
41871                  49.839                         1.5            67.92   
59539                  79.602                         1.3            74.25   
3935                      NaN                         NaN            76.29   
57045                     NaN                         NaN              NaN   
75170                  47.953                         0.7            65.46   
28257                     NaN                         6.3            66.47   
80366                     NaN                         2.8            77.91   
21298                     NaN                         1.4            67.11   
83861                  36.791                         1.7            61.49   

       hu

In [6]:
# Array slicing also works - this shows rows 40 up to (not including) 45.
data[40:45]

iso_code      continent   location       date  total_cases  new_cases  \
48734      MEX  North America     Mexico 2020-01-13          NaN        NaN   
59547      PER  South America       Peru 2020-01-13          NaN        NaN   
75180      THA           Asia   Thailand 2020-01-13          NaN        NaN   
3021       ARG  South America  Argentina 2020-01-13          NaN        NaN   
59548      PER  South America       Peru 2020-01-14          NaN        NaN   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
48734                 NaN           NaN         NaN                  NaN  ...   
59547                 NaN           NaN         NaN                  NaN  ...   
75180                 NaN           NaN         NaN                  NaN  ...   
3021                  NaN           NaN         NaN                  NaN  ...   
59548                 NaN           NaN         NaN                  NaN  ...   

       gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
48734       17336.469              2.5                152.783   
59547       12236.706              3.5                 85.755   
75180       16277.671              0.1                109.861   
3021        18933.907              0.6                191.032   
59548       12236.706              3.5                 85.755   

       diabetes_prevalence  female_smokers  male_smokers  \
48734                13.06             6.9          21.4   
59547                 5.95             4.8           NaN   
75180                 7.04             1.9          38.8   
3021                  5.50            16.2          27.7   
59548                 5.95             4.8           NaN   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
48734                  87.847                        1.38            75.05   
59547                     NaN                        1.60            76.74   
75180                  90.670                        2.10            77.15   
3021                      NaN                        5.00            76.67   
59548                     NaN                        1.60            76.74   

       human_development_index  
48734                    0.779  
59547                    0.777  
75180                    0.777  
3021                     0.845  
59548                    0.777  

[5 rows x 59 columns]